python/uv/private/toolchains_repo.bzl

```
@@ -0,0 +1,50 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"Creates a repository to hold toolchains"
+
+_TOOLCHAIN_TEMPLATE = """
+toolchain(
+    name = "{name}",
+    target_compatible_with = {compatible_with},
+    toolchain = "{toolchain_label}",
+    toolchain_type = "{toolchain_type}",
+)
+"""
+
+def _toolchains_repo_impl(repository_ctx):
+    build_content = ""
+    for toolchain_name in repository_ctx.attr.toolchain_names:
+        toolchain_label = repository_ctx.attr.toolchain_labels[toolchain_name]
+        toolchain_compatible_with = repository_ctx.attr.toolchain_compatible_with[toolchain_name]
+
+        build_content += _TOOLCHAIN_TEMPLATE.format(
+            name = toolchain_name,
+            toolchain_type = repository_ctx.attr.toolchain_type,
+            toolchain_label = toolchain_label,
+            compatible_with = json.encode(toolchain_compatible_with),
```

An alternative expression could be: That will give you the right pretty printed text.
```suggestion
            compatible_with = render.list(toolchain_compatible_with),
```

---

Addressed in: https://github.com/bazelbuild/rules_python/pull/2062